src/array/cpu/spmm_blocking_libxsmm.h

```diff
@@ -0,0 +1,598 @@
+ * \param is_cmp Is the reduction operation a compare operation.
+ * \note libxsmm_dispatch_meltw_opreduce_vecs_idx creates a JIT'ed kernel.
+ *       Given a node u, the kernel performs an elementwise "Op" on the
+ *       features of the neighbors and/or the edges incident on u.
+ *       Subsequently, it performs an elementwise "Redop" on all such
+ *       features created and stores into the feature of node u.
+ *       It uses a SIMD and a cache efficient design and also provides
+ *       support to enable software prefetching if needed. For IdType,
+ *       it supports INT32 and INT64. For DType, it supports BF16 and FP32.
+ *       It supports all the "Ops" and "Redops" supported by DGL. Once a
+ *       kernel is generated by libxsmm_dispatch_meltw_opreduce_vecs_idx,
+ *       it is cached for the entire duration of the execution of a program
+ *       so that subsequently if the kernel is needed again, it just returns
+ *       the cached copy.
+ */
+template <typename IdType, typename DType, typename Op>
+inline libxsmm_meltwfunction_opreduce_vecs_idx SpMMCreateLibxsmmKernel(
+    bool has_idx,
+    IdType N,
+    libxsmm_meltw_opreduce_vecs_flags redop_flag,
+    bool is_cmp) {
+  int _ld = N;
+  libxsmm_meltw_opreduce_vecs_flags opredop_flags;
+  // First, set the Op in the opredop_flags
+  if (std::is_same<Op, op::Add<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_ADD;
+  } else if (std::is_same<Op, op::Sub<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_SUB;
+  } else if (std::is_same<Op, op::Mul<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_MUL;
+  } else if (std::is_same<Op, op::Div<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_DIV;
+  } else if (std::is_same<Op, op::CopyLhs<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_COPY;
+  } else if (std::is_same<Op, op::CopyRhs<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_COPY;
+  }
+  // Second, set which of lhs or rhs is considered first and second operand.
+  // This is needed since libxsmm assumes that the copy operation always copies the first operand.
+  // So, if we need to copy rhs, we need to set that as the first operand.
+  // For rhs, we also set whether to use implicit indices or provided indices.
+  if (std::is_same<Op, op::CopyLhs<DType>>::value) {
+    opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags |
+                     LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OPORDER_VECIDX_VECIN);
+  } else if (std::is_same<Op, op::CopyRhs<DType>>::value) {
+    opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags |
+                     LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OPORDER_VECIN_VECIDX);
+    if (!has_idx) {
+      opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags |
+                       LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_IMPLICIT_INDEXED_VECIDX);
```

Comment:
Why no need to set `LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_INDEXED_VEC` flag as in L238-241?

Reply:
That is how libxsmm API works. There is no need to set `LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_INDEXED_VEC` flag as in L238-241?